### PR TITLE
Read only transactions

### PR DESCRIPTION
**Pull request by geoffjohn11**

### CODE REVIEW

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:28]: Any reason this one is abstract? It should be implemented like the others, no?

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:116]: I would mark those a `override final val`

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:133]: I would explicitly `override final` those

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:138]: I believe you can actually override this and make the return type precise to this specific subclass.

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:147]: You should not need to implement those, since those should come from the parent trait.

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:173]: Ditto all the previous comments.

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:282]: I would not complicate this that much with that `f` function, but rather just make this method return the triplet `(String, Map[String, QueryParam], List[Int])`

**Comment by geoffjohn11** [comment]
> Hi @geoffjohn11 thanks for looking into this. Even if I added a lot of comments, most are very minor so great job!
> 
> Quick question, what do you think of using `query` & `readQuery` instead of `command` & `query`?

Thanks for the review!  It's awesome how fast you are to respond!  I think query and readQuery sound good, I'll make that change.

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:25]: This could be a `final def`

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:113]: This must be `protected`

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:140]: `override final val`

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:169]: `override final val`

File: core/src/main/scala/neotypes/DeferredQuery.scala
Comment (BalmungSan) [inline core/src/main/scala/neotypes/DeferredQuery.scala:171]: `override def`

**Review by BalmungSan — COMMENTED** [review]
Hi @geoffjohn11 thanks for looking into most of the comments, only a few remain but most of them are just minor style things that I could just fix before or after merging this.

However, there is still one big issue with the **Scaladoc**, we can't lose all that.
Could you please look into that? Feel free to ping me on **gitter** if you have questions about that.

**Comment by BalmungSan** [comment]
Hi @geoffjohn11 sorry for highjacking your PR, I cloned it to apply the minor style recommendations since I didn't want to block the PR because of that... and then I got a little bit carried over 😅 

I found that the previous implementation was a bit misleading since it would allow running write operations on a `readOnlyQuery` , also `collectAs` was not using the `readOnlyTransaction`
While fixing that I also found an opportunity to refactor how the interpolation of `DefferedQueries` worked; but I am not totally sure if I broke something.

Sadly my changes introduced a lot of errors in the tests and I don't have more free time, if you want and can take a look at those I would be very grateful.
Otherwise, I may resume my work here next week, so don't worry. I broke it and I will fix it.

Once the tests are green again all that would be missing would be complement the docs in the microsite to mention the `readOnly` variants and then we can merge this!

**Comment by geoffjohn11** [comment]
> Hi @geoffjohn11 sorry for highjacking your PR, I cloned it to apply the minor style recommendations since I didn't want to block the PR because of that... and then I got a little bit carried over sweat_smile

No problem!  Thanks for helping!
> 
> I found that the previous implementation was a bit misleading since it would allow running write operations on a `readOnlyQuery` , also `collectAs` was not using the `readOnlyTransaction` While fixing that I also found an opportunity to refactor how the interpolation of `DefferedQueries` worked; but I am not totally sure if I broke something.
> 
> Sadly my changes introduced a lot of errors in the tests and I don't have more free time, if you want and can take a look at those I would be very grateful. Otherwise, I may resume my work here next week, so don't worry. I broke it and I will fix it.

I'll see if I can take a look!
> 
> Once the tests are green again all that would be missing would be complement the docs in the microsite to mention the `readOnly` variants and then we can merge this!

**Review by BalmungSan — APPROVED** [review]
Hi @geoffjohn11 thanks for helping with the tests, I removed the ones that were still failing since they were not doing anything at all now and the underlying logic was already tested somewhere else.

I also give a quick look at the docs but nothing fancy. If you want to add something more comprehensive it would be great.

BTW, since I also made a lot of changes I would like it if you can give a final review of the PR.
**Github** won't let you approve it since it's your own PR but you can let me through a comment here or in **gitter**.

**Pull request merged**
